### PR TITLE
Add initial implementation of qwen3.

### PR DESCRIPTION
**Pull request by jenriver**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+   git remote add upstream https://www.github.com/jax-ml/bonsai
+   ```
+
+5. Create a branch where you will develop from:
+
+   ```bash
+   git checkout -b name-of-change
+   ```
+
+   And implement your changes using your favorite editor (we recommend
```
[the remark this answers is not in the document]
Comment (jenriver) [reply CONTRIBUTING.md:61]: Thanks, applied.

File: bonsai/generate/utils.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+
+def next_power_of_2(x: int) -> int:
+    """Returns the next power of 2 that is not smaller than x."""
+    if x == 0:
+        return 1
+    return int(2 ** int(jnp.ceil(jnp.log2(x))))
+
+
+def pad_to_length(
```
[the remark this answers is not in the document]
Comment (jenriver) [reply bonsai/generate/utils.py:71]: Right. Omitting, and all pad references in current code uses `jnp.pad`.

File: README.md [hunk 3]
```diff
@@ -1,3 +1,56 @@
-# bonsai
+#  (Work-in-progress) Bonsai
 
-(Work-in-progress) Future collection of JAX model implementations.
+[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](LICENSE)
+
+A collection of State-of-the-Art models implemented using JAX [NNX](https://flax.readthedocs.io/en/v0.8.3/experimental/nnx/index.html). This repository serves as a curated list of well-documented and easy-to-understand JAX implementations of common machine learning models.
```
Comment (skye) [inline README.md:5]: Do we wanna say State of the Art? Like what does that mean? Maybe just remove "A collection of models..."

File: bonsai/models/qwen3/tests/run_model.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import jax
+import jax.numpy as jnp
+import numpy as np
+from flax import nnx
+from huggingface_hub import snapshot_download
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from bonsai.generate import sampler, utils
+from bonsai.models.qwen3 import model, params
+
```
Comment (skye) [inline bonsai/models/qwen3/tests/run_model.py:13]: This is ok for now, but there should be actual unit tests for this model (i.e. that can be run with pytest) to check numerical accuracy against a baseline implemention, including logits. This might be good inspiration: https://github.com/jax-ml/jax-llm-examples/blob/main/deepseek_r1_jax/tests/test_numerics.py

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+.ipynb_checkpoints
+.DS_Store
+.mypy_cache/
+.pytype/
+.idea
+.vscode
+.envrc
+
+# virtualenv/venv directories
+/venv/
```
Comment (skye) [inline .gitignore:15]: Are these absolute paths? Like when would folders like bin, include, etc. show up in this git repo?

File: bonsai/generate/utils.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+def find_first_non_pad_idx(ids, pad_id):
+    """Finds the index of the first non-pad token."""
+    mask = ids != pad_id
+    if jnp.any(mask):
+        return jnp.argmax(mask)
+    else:
+        return 0
+
+
+def find_first_eos_idx(ids, eos_id):
```
[the remark this answers is not in the document]
Comment (jenriver) [reply bonsai/generate/utils.py:115]: For this, we do want to return ids.shape[0] (not 0) if there is no eos token, which is not easily expressible in one line jnp unless we do something like below. wdyt?

```
end_idx = jnp.where(
    jnp.any(token_buffer[max_prompt_length:] == tokenizer.eos_token_id),
    jnp.argmax(token_buffer[max_prompt_length:] == tokenizer.eos_token_id),
    token_buffer[max_prompt_length:].shape[0]
) + max_prompt_length
```

File: bonsai/generate/utils.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    attn_mask = input_mask[..., None, :]
+    causal_mask = jnp.tril(jnp.ones((seq_len, seq_len), dtype=jnp.bool_))
+    attn_mask *= causal_mask[None, ...]
+    padding = cache_size - seq_len
+    assert padding >= 0
+    attn_mask = jnp.pad(attn_mask, (*((0, 0) for _ in range(attn_mask.ndim - 1)), (0, padding)))
+    return attn_mask
+
+
+def next_power_of_2(x: int) -> int:
```
[the remark this answers is not in the document]
Comment (jenriver) [reply bonsai/generate/utils.py:64]: Done. Added comment for clarity.

File: bonsai/generate/utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jenriver) [reply bonsai/generate/utils.py:64]: done

File: bonsai/generate/sampler.py [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+        done=done,
+        total_sampling_steps=sampler_state.total_sampling_steps,
+        temperature=sampler_state.temperature,
+        sampling_parameters=sampler_state.sampling_parameters,
+        seed=sampler_state.seed,
+        sampling_mode=sampler_state.sampling_mode,
+    )
+
+
+@flax.struct.dataclass
```
[the remark this answers is not in the document]
Comment (jenriver) [reply bonsai/generate/sampler.py]: sampler removed with simpler API. resolving.

File: bonsai/models/qwen3/tests/run_model.py (hunk 4 above)
Comment (jenriver) [reply bonsai/models/qwen3/tests/run_model.py:14]: ack, will do in follow up cl

File: bonsai/models/qwen3/tests/run_model.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    if i == 1:  # Avoid XLA warmup time when measuring trace.
+        t = time.perf_counter()
+        jax.profiler.start_trace("/tmp/profile-data")
+    input_mask = jnp.array([next_token]) == tokenizer.pad_token_id
+    attention_mask = utils.compute_attention_masks(token_len + i, cache[0].size, input_mask)
+    positions = jnp.array([token_len + i])
+    logits, state = modeling.forward(graphdef, state, [next_token], positions, attention_mask)
+    next_token = jnp.argmax(logits[:, -1], axis=-1, keepdims=True)[0]  # greedy sampling
+    tokens_list.append(next_token)
+jax.profiler.stop_trace()
```
[the remark this answers is not in the document]
Comment (jenriver) [reply bonsai/models/qwen3/tests/run_model.py:67]: Is this necessary? I could do something like this, but this results in a much slower performance because `jnp.append` is very expensive.


```
# decode
tokens_list = jnp.array(next_token)
# Run `xprof --logdir /tmp/profile-data/` to see the program traces.
for i in range(20):
    if i == 1:  # Avoid XLA warmup time when measuring trace.
        t = time.perf_counter()
        jax.profiler.start_trace("/tmp/profile-data")
    input_mask = jnp.array([next_token]) == tokenizer.pad_token_id
    attention_mask = utils.compute_attention_masks(token_len + i, cache[0].size, input_mask)
    positions = jnp.array([token_len + i])
    logits, state = modeling.forward(graphdef, state, [next_token], positions, attention_mask)
    next_token = jnp.argmax(logits[:, -1], axis=-1, keepdims=True)[0]  # greedy sampling
    tokens_list = jnp.append(tokens_list, next_token)
tokens_list.block_until_ready()
jax.profiler.stop_trace()
```

[the remark this answers is not in the document]
Comment (rdyro) [reply bonsai/models/qwen3/tests/run_model.py:67]: regarding the mask:
```
    input_mask = jnp.array([next_token]) == tokenizer.pad_token_id
    attention_mask = utils.compute_attention_masks(token_len + i, cache[0].size, input_mask)
    positions = jnp.array([token_len + i])

```
these should be under forward to minimize separate jit calls (and avoid fully realizing the mask which can be very expensive in longer context lengths)

File: bonsai/models/qwen3/tests/run_model.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+# decode
+tokens_list = [next_token]
+# Run `xprof --logdir /tmp/profile-data/` to see the program traces.
+for i in range(20):
+    if i == 1:  # Avoid XLA warmup time when measuring trace.
+        t = time.perf_counter()
+        jax.profiler.start_trace("/tmp/profile-data")
+    attention_mask = utils.make_attention_mask(
+        jnp.array([next_token]), cache[0].size, False, tokenizer.pad_token_id, token_len + i
```
[the remark this answers is not in the document]
Comment (jenriver) [reply bonsai/models/qwen3/tests/run_model.py:61]: No longer an issue with current implementation. next_tokens is much more performant when defined in jnp than np , which results in many dispersed sparse thread pools in this case.

File: bonsai/models/qwen3/tests/run_model.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    if i == 5:  # profile steps 1-5
+        jax.block_until_ready(tokens_list)
+        jax.profiler.stop_trace()
+        t = time.perf_counter()  # measure steps 6-10
+jax.block_until_ready(tokens_list)
+print(f"{time.perf_counter() - t:.4f} s")
+tokens_list = jnp.squeeze(jnp.array(tokens_list), axis=2).T
+for i, q in enumerate(query):
+    print(f"User:\n {q}")
+    print(f"Answer:\n {''.join([tokenizer.decode(t) for t in tokens_list[i]])}\n\n")
```
Comment (rdyro) [inline bonsai/models/qwen3/tests/run_model.py:66]: nit: tokens_list.ndim == 2, so ideally two indices to slice into it `tokens_list[i, :]`

File: bonsai/models/qwen3/tests/run_model.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        jax.profiler.start_trace("/tmp/profile-data")
+    next_tokens, state = modeling.forward(graphdef, state, next_tokens, False, token_len + i)
+    tokens_list.append(next_tokens)
+    if i == 5:  # profile steps 1-5
+        jax.block_until_ready(tokens_list)
+        jax.profiler.stop_trace()
+        t = time.perf_counter()  # measure steps 6-10
+jax.block_until_ready(tokens_list)
+print(f"{time.perf_counter() - t:.4f} s")
+tokens_list = jnp.squeeze(jnp.array(tokens_list), axis=2).T
```
Comment (rdyro) [inline bonsai/models/qwen3/tests/run_model.py:63]: nit: a little awkward, is this `jnp.concatenate(tokens_list, -1).T` ?

File: bonsai/models/qwen3/modeling.py [hunk 13]
```diff
@@ -0,0 +1,472 @@
...
+@partial(jax.jit, donate_argnums=(1), static_argnums=(3))
+def forward(
+    graphdef: nnx.GraphDef[tuple[nnx.Module, Cache]],
+    state: nnx.State,
+    input_tokens: jaxtyping.Array,  # [B, L]
+    is_prefill: bool,
+    time_step: int = 0,
+) -> tuple[jaxtyping.Array, Cache | None]:
+    _, query_len = input_tokens.shape
+    model, cache = nnx.merge(graphdef, state)
```
Comment (rdyro) [inline bonsai/models/qwen3/modeling.py:453]: state comprises both the model and the cache, that seems confusing, can we separate those please?

File: bonsai/models/qwen3/modeling.py [hunk 14]
```diff
@@ -0,0 +1,472 @@
...
+            einsum_str="BTD,DV->BTV", shape=(config.embed_dim, config.vocab_size), rngs=rngs, sharding=shd_config.emb_dv
+        )
+
+
+@partial(jax.jit, donate_argnums=(1), static_argnums=(3))
+def forward(
+    graphdef: nnx.GraphDef[tuple[nnx.Module, Cache]],
+    state: nnx.State,
+    input_tokens: jaxtyping.Array,  # [B, L]
+    is_prefill: bool,
```
Comment (rdyro) [inline bonsai/models/qwen3/modeling.py:449]: combination of `is_prefill` and `time_step` is not ideal (e.g., how do we count chunked prefill)

can you take a look at how positions and offset is calculated from the cache state and segment_ids in [https://github.com/jax-ml/jax-llm-examples/blob/99db94d32a8357cc7b7dfd82d045999c823d870f/llama3/llama3_jax/model.py#L772](https://github.com/jax-ml/jax-llm-examples/blob/99db94d32a8357cc7b7dfd82d045999c823d870f/llama3/llama3_jax/model.py#L772)

File: bonsai/models/qwen3/modeling.py (hunk 13 above)
Comment (jenriver) [reply bonsai/models/qwen3/modeling.py:445]: synced offline to address this later

**Review by rdyro — DISMISSED** [review]
Awesome work!
